File: R/binary_search.r
# Function for binary search
binarySearch <- function(vector, target) {
  left <- 1
  right <- length(vector)

  while (left <= right) {
    mid <- floor((left + right) / 2)

    if (vector[mid] == target) {
      return(mid)
    } else if (vector[mid] < target) {
      left <- mid + 1
    } else {
      right <- mid - 1
    }
  }

  return(-1) # Target not found
}

# Read command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Validate and parse arguments
if (length(args) >= 2) {
  vector <- as.integer(unlist(strsplit(args[1], ",")))
  target <- as.integer(args[2])

  # Perform binary search
  index <- binarySearch(vector, target)

  # Output the result
  if (index != -1) {
    cat("Target found at index:", index, "\n")
  } else {
    cat("Target not found in the vector.\n")
  }
} else {
  cat("Usage: Rscript binary_search.R \"1,3,5,7,9\" 5\n")
}

File: cpp/binary_search.cpp
#include <iostream>
#include <vector>
#include <algorithm> // For std::sort

// Function for binary search
int binarySearch(const std::vector<int>& arr, int target) {
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target) {
            return mid; // Target found
        }

        if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return -1; // Target not found
}

int main() {
    int n, target;
    std::cout << "Enter number of elements: ";
    std::cin >> n;

    std::vector<int> arr(n);
    std::cout << "Enter " << n << " elements in sorted order:\n";
    for (int i = 0; i < n; ++i) {
        std::cin >> arr[i];
    }

    std::cout << "Enter the element to search for: ";
    std::cin >> target;

    // Optionally sort the array if the input is not guaranteed to be sorted
    // std::sort(arr.begin(), arr.end());

    int result = binarySearch(arr, target);

    if (result != -1) {
        std::cout << "Element found at index: " << result << std::endl;
    } else {
        std::cout << "Element not found in the array." << std::endl;
    }

    return 0;
}
